cirq-core/cirq/protocols/json_serialization.py

```diff
@@ -249,7 +269,24 @@ def dataclass_json_dict(obj: Any, namespace: str = None) -> Dict[str, Any]:
     Although not as elegant, you may want to consider explicitly defining `_json_dict_` on your
     dataclasses which simply `return dataclass_json_dict(self)`.
     """
-    return obj_to_dict_helper(obj, [f.name for f in dataclasses.fields(obj)], namespace=namespace)
+    attribute_names = [f.name for f in dataclasses.fields(obj)]
+    if namespace is not None:
+        return obj_to_dict_helper(obj, attribute_names, namespace=namespace)
+    else:
+        return _obj_to_dict_helper_helper(obj, attribute_names)
+
+
+def _json_dict_with_cirq_type(obj: Any):
+    base_dict = obj._json_dict_()
+    if 'cirq_type' in base_dict:
+        # TODO: upgrade to ValueError in v0.15
+        warnings.warn(
+            f"Found 'cirq_type': '{base_dict['cirq_type']}' in _json_dict_. "
+            f"Custom values of this field are not permitted, and will produce "
+            "an error starting in Cirq v0.15.",
+            DeprecationWarning,
+        )
+    return {'cirq_type': json_cirq_type(type(obj)), **base_dict}
```

Comment:
would this work with `json_cirq_type(obj)`? or does it need a type? pros and cons of handling this here vs in `json_cirq_type`?

Reply:
I'm pretty sure I could make `json_cirq_type` accept both objects and types - we'd just need to differentiate between the two inside the method. However, the only case I can think of where a user needs to call `json_cirq_type` directly is for type-serialization, in which case they ought to be passing in types anyways.